net/dns/resolver: data race

from https://github.com/tailscale/tailscale/pull/1655/checks?check_run_id=2280611645
```
==================
WARNING: DATA RACE
Write at 0x00c000147650 by goroutine 83:
  internal/race.Write()
      /opt/hostedtoolcache/go/1.16.3/x64/src/internal/race/race.go:41 +0x125
  sync.(*WaitGroup).Wait()
      /opt/hostedtoolcache/go/1.16.3/x64/src/sync/waitgroup.go:128 +0x126
  tailscale.com/net/dns/resolver.(*fwdConn).closeConnLocked()
      /home/runner/work/tailscale/tailscale/net/dns/resolver/forwarder.go:474 +0x114
  tailscale.com/net/dns/resolver.(*fwdConn).close()
      /home/runner/work/tailscale/tailscale/net/dns/resolver/forwarder.go:487 +0x12e
  tailscale.com/net/dns/resolver.(*forwarder).Close()
      /home/runner/work/tailscale/tailscale/net/dns/resolver/forwarder.go:164 +0x15d
  tailscale.com/net/dns/resolver.(*Resolver).Close()
      /home/runner/work/tailscale/tailscale/net/dns/resolver/tsdns.go:184 +0x148
  tailscale.com/net/dns/resolver.TestDelegateCollision()
      /home/runner/work/tailscale/tailscale/net/dns/resolver/tsdns_test.go:480 +0xd2e
  testing.tRunner()
      /opt/hostedtoolcache/go/1.16.3/x64/src/testing/testing.go:1193 +0x202

Previous read at 0x00c000147650 by goroutine 12:
  internal/race.Read()
      /opt/hostedtoolcache/go/1.16.3/x64/src/internal/race/race.go:37 +0x206
  sync.(*WaitGroup).Add()
      /opt/hostedtoolcache/go/1.16.3/x64/src/sync/waitgroup.go:71 +0x219
  tailscale.com/net/dns/resolver.(*fwdConn).send()
      /home/runner/work/tailscale/tailscale/net/dns/resolver/forwarder.go:357 +0x32d
  tailscale.com/net/dns/resolver.(*forwarder).send()
      /home/runner/work/tailscale/tailscale/net/dns/resolver/forwarder.go:188 +0x124
  tailscale.com/net/dns/resolver.(*forwarder).forward()
      /home/runner/work/tailscale/tailscale/net/dns/resolver/forwarder.go:294 +0x516
  tailscale.com/net/dns/resolver.(*Resolver).poll()
      /home/runner/work/tailscale/tailscale/net/dns/resolver/tsdns.go:328 +0x804

Goroutine 83 (running) created at:
  testing.(*T).Run()
      /opt/hostedtoolcache/go/1.16.3/x64/src/testing/testing.go:1238 +0x5d7
  testing.runTests.func1()
      /opt/hostedtoolcache/go/1.16.3/x64/src/testing/testing.go:1511 +0xa6
  testing.tRunner()
      /opt/hostedtoolcache/go/1.16.3/x64/src/testing/testing.go:1193 +0x202
  testing.runTests()
      /opt/hostedtoolcache/go/1.16.3/x64/src/testing/testing.go:1509 +0x612
  testing.(*M).Run()
      /opt/hostedtoolcache/go/1.16.3/x64/src/testing/testing.go:1417 +0x3b3
  main.main()
      _testmain.go:117 +0x356

Goroutine 12 (running) created at:
  tailscale.com/net/dns/resolver.New()
      /home/runner/work/tailscale/tailscale/net/dns/resolver/tsdns.go:118 +0x446
  tailscale.com/net/dns/resolver.TestDelegateCollision()
      /home/runner/work/tailscale/tailscale/net/dns/resolver/tsdns_test.go:416 +0x278
  testing.tRunner()
      /opt/hostedtoolcache/go/1.16.3/x64/src/testing/testing.go:1193 +0x202
==================
--- FAIL: TestDelegateCollision (0.01s)
```

--- a/net/dns/resolver/forwarder.go
+++ b/net/dns/resolver/forwarder.go
@@ -13,7 +13,6 @@ import (
 	"hash/crc32"
 	"math/rand"
 	"net"
-	"os"
 	"sync"
 	"time"
 
@@ -38,8 +37,6 @@ const (
 )
 
 var errNoUpstreams = errors.New("upstream nameservers not set")
-
-var aLongTimeAgo = time.Unix(0, 1)
 
 type forwardingRecord struct {
 	src       netaddr.IPPort
@@ -303,8 +300,6 @@ type fwdConn struct {
 	// logf allows a fwdConn to log.
 	logf logger.Logf
 
-	// wg tracks the number of outstanding conn.Read and conn.Write calls.
-	wg sync.WaitGroup
 	// change allows calls to read to block until a the network connection has been replaced.
 	change *sync.Cond
 
@@ -352,15 +347,12 @@ func (c *fwdConn) send(packet []byte, dst netaddr.IPPort) {
 		}
 		c.mu.Unlock()
 
-		a := dst.UDPAddr()
-		c.wg.Add(1)
-		_, err := conn.WriteTo(packet, a)
-		c.wg.Done()
+		_, err := conn.WriteTo(packet, dst.UDPAddr())
 		if err == nil {
 			// Success
 			return
 		}
-		if errors.Is(err, os.ErrDeadlineExceeded) {
+		if errors.Is(err, net.ErrClosed) {
 			// We intentionally closed this connection.
 			// It has been replaced by a new connection. Try again.
 			continue
@@ -429,14 +421,12 @@ func (c *fwdConn) read(out []byte) int {
 		}
 		c.mu.Unlock()
 
-		c.wg.Add(1)
 		n, _, err := conn.ReadFrom(out)
-		c.wg.Done()
 		if err == nil {
 			// Success.
 			return n
 		}
-		if errors.Is(err, os.ErrDeadlineExceeded) {
+		if errors.Is(err, net.ErrClosed) {
 			// We intentionally closed this connection.
 			// It has been replaced by a new connection. Try again.
 			continue
@@ -468,10 +458,7 @@ func (c *fwdConn) closeConnLocked() {
 	if c.conn == nil {
 		return
 	}
-	// Unblock all readers/writers, wait for them, close the connection.
-	c.conn.SetDeadline(aLongTimeAgo)
-	c.wg.Wait()
-	c.conn.Close()
+	c.conn.Close() // unblocks all readers/writers, they'll pick up the next connection.
 	c.conn = nil
 }
 

